File: examples/csharp/Customer/src/CustomerService.cs
using Angzarr;
using Examples;
using Grpc.Core;
using Serilog;

namespace Angzarr.Examples.Customer;

/// <summary>
/// gRPC service implementation for customer business logic.
/// </summary>
public class CustomerService : BusinessLogic.BusinessLogicBase
{
    private readonly ICustomerLogic _logic;
    private readonly Serilog.ILogger _logger;

    public CustomerService(ICustomerLogic logic, Serilog.ILogger logger)
    {
        _logic = logic;
        _logger = logger.ForContext<CustomerService>();
    }

    public override Task<BusinessResponse> Handle(ContextualCommand request, ServerCallContext context)
    {
        try
        {
            var events = ProcessCommand(request);
            var response = new BusinessResponse { Events = events };
            return Task.FromResult(response);
        }
        catch (CommandValidationException ex)
        {
            throw new RpcException(new Status(ex.StatusCode, ex.Message));
        }
        catch (Exception ex)
        {
            _logger.Error(ex, "Unexpected error processing command");
            throw new RpcException(new Status(StatusCode.Internal, $"Internal error: {ex.Message}"));
        }
    }

    private EventBook ProcessCommand(ContextualCommand request)
    {
        var cmdBook = request.Command;
        var priorEvents = request.Events;

        if (cmdBook == null || cmdBook.Pages.Count == 0)
            throw CommandValidationException.InvalidArgument("CommandBook has no pages");

        var cmdPage = cmdBook.Pages[0];
        if (cmdPage.Command == null)
            throw CommandValidationException.InvalidArgument("Command page has no command");

        var state = _logic.RebuildState(priorEvents);
        var command = cmdPage.Command;
        var typeUrl = command.TypeUrl;

        EventBook result;

        if (typeUrl.EndsWith("CreateCustomer"))
        {
            var cmd = command.Unpack<CreateCustomer>();
            result = _logic.HandleCreateCustomer(state, cmd.Name, cmd.Email);
        }
        else if (typeUrl.EndsWith("AddLoyaltyPoints"))
        {
            var cmd = command.Unpack<AddLoyaltyPoints>();
            result = _logic.HandleAddLoyaltyPoints(state, cmd.Points, cmd.Reason);
        }
        else if (typeUrl.EndsWith("RedeemLoyaltyPoints"))
        {
            var cmd = command.Unpack<RedeemLoyaltyPoints>();
            result = _logic.HandleRedeemLoyaltyPoints(state, cmd.Points, cmd.RedemptionType);
        }
        else
        {
            throw CommandValidationException.InvalidArgument($"Unknown command type: {typeUrl}");
        }

        // Add cover from command book
        result.Cover = cmdBook.Cover;
        return result;
    }
}

File: examples/cpp/customer/src/customer_service.cc
#include "customer_logic.hpp"
#include "logging.hpp"
#include "validation_error.hpp"
#include "angzarr.grpc.pb.h"
#include <grpcpp/grpcpp.h>

namespace customer {

class CustomerService final : public angzarr::BusinessLogic::Service {
public:
    grpc::Status Handle(grpc::ServerContext* context,
                        const angzarr::ContextualCommand* request,
                        angzarr::BusinessResponse* response) override {
        try {
            const auto& cmd_book = request->command();
            const auto* prior_events = request->has_events() ? &request->events() : nullptr;

            if (cmd_book.pages().empty()) {
                return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT, "CommandBook has no pages");
            }

            const auto& cmd_page = cmd_book.pages(0);
            if (!cmd_page.has_command()) {
                return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT, "Command page has no command");
            }

            auto state = CustomerLogic::rebuild_state(prior_events);
            const auto& type_url = cmd_page.command().type_url();

            google::protobuf::Any event_any;

            if (type_url.find("CreateCustomer") != std::string::npos) {
                examples::CreateCustomer cmd;
                cmd_page.command().UnpackTo(&cmd);
                angzarr::log_info("customer", "creating_customer",
                    {{"name", cmd.name()}, {"email", cmd.email()}});
                auto event = CustomerLogic::handle_create_customer(state, cmd.name(), cmd.email());
                event_any.PackFrom(event);
            } else if (type_url.find("AddLoyaltyPoints") != std::string::npos) {
                examples::AddLoyaltyPoints cmd;
                cmd_page.command().UnpackTo(&cmd);
                angzarr::log_info("customer", "adding_loyalty_points",
                    {{"points", cmd.points()}, {"reason", cmd.reason()}});
                auto event = CustomerLogic::handle_add_loyalty_points(state, cmd.points(), cmd.reason());
                event_any.PackFrom(event);
            } else if (type_url.find("RedeemLoyaltyPoints") != std::string::npos) {
                examples::RedeemLoyaltyPoints cmd;
                cmd_page.command().UnpackTo(&cmd);
                angzarr::log_info("customer", "redeeming_loyalty_points",
                    {{"points", cmd.points()}, {"redemption_type", cmd.redemption_type()}});
                auto event = CustomerLogic::handle_redeem_loyalty_points(
                    state, cmd.points(), cmd.redemption_type());
                event_any.PackFrom(event);
            } else {
                return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT,
                    "Unknown command type: " + type_url);
            }

            auto* event_book = response->mutable_events();
            if (cmd_book.has_cover()) {
                *event_book->mutable_cover() = cmd_book.cover();
            }

            auto* event_page = event_book->add_pages();
            event_page->set_num(0);
            *event_page->mutable_event() = event_any;
            event_page->mutable_created_at()->set_seconds(
                std::chrono::system_clock::to_time_t(std::chrono::system_clock::now()));

            return grpc::Status::OK;

        } catch (const angzarr::ValidationError& e) {
            return e.to_grpc_status();
        }
    }
};

std::unique_ptr<angzarr::BusinessLogic::Service> create_customer_service() {
    return std::make_unique<CustomerService>();
}

}  // namespace customer
